presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java

```diff
@@ -100,49 +136,117 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
                     hiveStorageTimeZone
             );
             if (pageSource.isPresent()) {
-                return pageSource.get();
+                return Optional.of(pageSource.get());
             }
         }
 
-        HiveRecordCursor recordCursor = getHiveRecordCursor(clientId, session, configuration, path, start, length, schema, effectivePredicate, partitionKeys, hiveColumns);
-        if (recordCursor != null) {
-            List<Type> columnTypes = ImmutableList.copyOf(transform(hiveColumns, input -> typeManager.getType(input.getTypeSignature())));
-            return new RecordPageSource(columnTypes, recordCursor);
-        }
-
-        throw new RuntimeException("Could not find a file reader for split " + hiveSplit);
-    }
-
-    protected HiveRecordCursor getHiveRecordCursor(
-            String clientId,
-            ConnectorSession session,
-            Configuration configuration,
-            Path path,
-            long start,
-            long length,
-            Properties schema,
-            TupleDomain<HiveColumnHandle> effectivePredicate,
-            List<HivePartitionKey> partitionKeys,
-            List<HiveColumnHandle> hiveColumns)
-    {
         for (HiveRecordCursorProvider provider : cursorProviders) {
-            Optional<HiveRecordCursor> cursor = provider.createHiveRecordCursor(
+            Optional<RecordCursor> cursor = provider.createRecordCursor(
                     clientId,
                     configuration,
                     session,
                     path,
                     start,
                     length,
                     schema,
-                    hiveColumns,
-                    partitionKeys,
+                    regularColumns,
                     effectivePredicate,
                     hiveStorageTimeZone,
                     typeManager);
+
             if (cursor.isPresent()) {
-                return cursor.get();
+                HiveRecordCursor hiveRecordCursor = new HiveRecordCursor(
+                        columnMappings,
+                        hiveStorageTimeZone,
+                        typeManager,
+                        cursor.get());
+                List<Type> columnTypes = ImmutableList.copyOf(transform(hiveColumns, input -> typeManager.getType(input.getTypeSignature())));
```

Comment:
We don't use `transform` any more. Rewrite using streaming API. Don't worry about Immutable.